lib/download.go

```diff
@@ -9,11 +9,86 @@ import (
 	"strings"
 )
 
-// DownloadFromURL : Downloads the binary from the source url
-func DownloadFromURL(installLocation string, url string) (string, error) {
+// DownloadFromURL : Downloads the terraform binary and its hash from the source url
+func DownloadFromURL(installLocation string, mirrorURL string, tfversion string, versionPrefix string, goos string, goarch string) (string, error) {
+	pubKeyFilename := filepath.Join(installLocation, "/", PubKeyPrefix+PubKeyId+pubKeySuffix)
+	zipUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_" + goos + "_" + goarch + ".zip"
+	hashUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS"
+	hashSignatureUrl := mirrorURL + tfversion + "/" + versionPrefix + tfversion + "_SHA256SUMS." + PubKeyId + ".sig"
+
+	err := downloadPublicKey(installLocation, pubKeyFilename)
+	if err != nil {
+		logger.Error("Could not download public PGP key file.")
+		return "", err
+	}
+
+	logger.Infof("Downloading %q", zipUrl)
+	zipFilePath, err := downloadFromURL(installLocation, zipUrl)
+	if err != nil {
+		logger.Error("Could not download zip file.")
+		return "", err
+	}
+
+	logger.Infof("Downloading %q", hashUrl)
+	hashFilePath, err := downloadFromURL(installLocation, hashUrl)
+	if err != nil {
+		logger.Error("Could not download hash file.")
+		return "", err
+	}
+
+	logger.Infof("Downloading %q", hashSignatureUrl)
+	hashSigFilePath, err := downloadFromURL(installLocation, hashSignatureUrl)
+	if err != nil {
+		logger.Error("Could not download hash signature file.")
+		return "", err
+	}
+
+	publicKeyFile, err := os.Open(pubKeyFilename)
+	if err != nil {
+		logger.Errorf("Could not open public key %q: %v", pubKeyFilename, err)
+		return "", err
+	}
+
+	signatureFile, err := os.Open(hashSigFilePath)
+	if err != nil {
+		logger.Errorf("Could not open hash signature file %q: %v", hashSigFilePath, err)
+		return "", err
+	}
+
+	targetFile, err := os.Open(zipFilePath)
+	if err != nil {
+		logger.Errorf("Could not open zip file %q: %v", zipFilePath, err)
+		return "", err
+	}
+
+	hashFile, err := os.Open(hashFilePath)
+	if err != nil {
+		logger.Errorf("Could not open hash file %q: %v", hashFilePath, err)
+		return "", err
+	}
+
+	var filesToCleanup []string
+	filesToCleanup = append(filesToCleanup, hashFilePath)
+	filesToCleanup = append(filesToCleanup, hashSigFilePath)
+
+	verified := checkSignatureOfChecksums(publicKeyFile, hashFile, signatureFile)
+	if !verified {
+		cleanup(filesToCleanup)
+		return "", errors.New("Signature of checksum file could not be verified")
+	}
+	match := checkChecksumMatches(hashFilePath, targetFile)
+	if !match {
+		cleanup(filesToCleanup)
+		return "", errors.New("Checksums did not match")
+	}
+	cleanup(filesToCleanup)
+	return zipFilePath, err
```

Comment:
We probably can do something like the below instead of adding `cleanup(filesToCleanup)` three times? (I'm not very familiar with Go and hence might be mixing up things)
```suggestion
    defer cleanup(filesToCleanup)

	verified := checkSignatureOfChecksums(publicKeyFile, hashFile, signatureFile)
	if !verified {
		return "", errors.New("Signature of checksum file could not be verified")
	}

	match := checkChecksumMatches(hashFilePath, targetFile)
	if !match {
		return "", errors.New("Checksums did not match")
	}

	return zipFilePath, err
```

Reply:
Will call defer cleanup which will automatically called on return